### PR TITLE
Dev tools: Improve developer UX for workflow building

### PR DESCRIPTION
**Pull request by thomas-**
This utilises the recently added O.currentAuthenticatedUser and enhancements to Impersonation endpoints (eg: redirects, impersonations while impersonating) to add tooling around workflow development:

* One click-impersonate: Quickly impersonate the current actionable user of a workflow from the action panel and return to the current page. (Where the actionable is a group, we get the first member of the group)
* Entity impersonate: Quickly impersonate any resolvable entity defined as part of a workflow with a sidebar on the right.
* Previous state: Quickly toggle back to the previous state from the workflow action panel

The tools are disabled by default but can be toggled on/off by a super user for their account from the 'Workflow admin' section.

We utilise the O.currentAuthenticatedUser API to allow debug/admin actions to be undertaken while impersonating if this is a development system (with plugin debugging enabled) and the user has enabled the tooling.

### CODE REVIEW

File: std_workflow_dev/js/std_workflow_debug.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            var row = select.length > 1 ? select[1] : undefined;
+            if(!row) { O.stop(); }
+            M._forceMoveToStateFromTimelineEntry(row, null);
+            return E.response.redirect(M.url);
+        });
+
+        // --------------------------------------------------------------------------
+
+        P.respond("GET,POST", "/do/workflow-dev/debug/debug-mode", [
+            {pathElement:0, as:"string"},
```
Comment (bensummers) [inline std_workflow_dev/js/std_workflow_debug.js]: Isn't this redundant now?

File: std_document_store_dev/requirements.schema [hunk 2]
```diff
@@ -0,0 +1,3 @@
+
+
+
```
Comment (bensummers) [inline std_document_store_dev/requirements.schema]: empty file!

File: std_workflow_dev/plugin.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+  "pluginName": "std_workflow_dev",
+  "pluginAuthor": "Haplo Services",
+  "pluginVersion": 1,
+  "displayName": "Workflow Development",
+  "displayDescription": "Development tools for workflow plugin",
+  "apiVersion": 4,
+  "depend": ["std_workflow"],
+  "privilegesRequired": ["pDatabase"],
```
Comment (bensummers) [inline std_workflow_dev/plugin.json]: Don't think it needs pDatabase?

File: std_document_store_dev/plugin.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "pluginName": "std_document_store_dev",
+  "pluginAuthor": "Haplo Services",
+  "pluginVersion": 1,
+  "displayName": "Document Store Development",
+  "displayDescription": "Development tools for document store plugin",
+  "installSecret": "02b34388d0f255452e4dbfa4ddea3946bd0b3205",
```
Comment (bensummers) [inline std_document_store_dev/plugin.json]: Standard plugins shouldn't have install secrets.

File: std_document_store_dev/js/docstore_workflow_debug.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+};
+
+if(O.PLUGIN_DEBUGGING_ENABLED) {
+
+    // UI for showing/hiding implemented by std_workflow_dev
+    var showDebugTools = function() {
+        return (O.PLUGIN_DEBUGGING_ENABLED &&
+            O.currentAuthenticatedUser &&
+            O.currentAuthenticatedUser.isSuperUser &&
+            O.currentAuthenticatedUser.id === 3 && // SUPPORT only for now
```
Comment (bensummers) [inline std_document_store_dev/js/docstore_workflow_debug.js]: Perhaps we should just let anyone use this? Some developers aren't going to be logging in via a dev portal which gets them SUPPORT.

File: std_document_store_dev/js/docstore_workflow_debug.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
+/* Haplo Platform                                     http://haplo.org
+ * (c) Haplo Services Ltd 2006 - 2016    http://www.haplo-services.com
```
Comment (bensummers) [inline std_document_store_dev/js/docstore_workflow_debug.js]: It's 2017 & https: now...

File: std_workflow_dev/js/std_workflow_debug.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    }
+};
+
+if(O.PLUGIN_DEBUGGING_ENABLED) {
+
+    P.workflow.registerOnLoadCallback(function(workflows) {
+        var showDebugTools = function() {
+            return (O.PLUGIN_DEBUGGING_ENABLED &&
+                O.currentAuthenticatedUser &&
+                O.currentAuthenticatedUser.isSuperUser &&
```
Comment (bensummers) [inline std_workflow_dev/js/std_workflow_debug.js]: This stuff is actually probably useful for support. Maybe these things do belong in the standard plugin?

File: std_workflow_dev/js/std_workflow_debug.js (hunk 1 above)
Comment (thomas-) [reply std_workflow_dev/js/std_workflow_debug.js]: Can be used as non-auth user is the only main difference currently I think, *but* I want to/am imagining that the two diverge, where debug has more power and abilities that wouldn't be suitable for administrators even in the production system.

File: std_document_store_dev/js/docstore_workflow_debug.js (hunk 5 above)
Comment (thomas-) [reply std_document_store_dev/js/docstore_workflow_debug.js]: Agreed. My intent here was being overly cautious until it's been used a bit and then dropping such a restriction.

File: std_workflow_dev/js/std_workflow_debug.js (hunk 7 above)
Comment (thomas-) [reply std_workflow_dev/js/std_workflow_debug.js]: Do you mean showing of admin tools while impersonating? Main worry for me is confusing users but maybe one for further down the line?

**Comment by thomas-** [comment]
fixed the obvious issues that needed tidying, and added a commit that modifies std_workflow slightly so that we get a more usable admin tools when debugging

**Pull request merged**